### PR TITLE
Offer to unban user during invite if inviter has sufficient permissions

### PR DESCRIPTION
**Pull request by t3chguy**
Fixes https://github.com/vector-im/element-web/issues/3222

image

---
Here's what your changelog entry will look like:

## ✨ Features
 * Offer to unban user during invite if inviter has sufficient permissions ([\#11256](https://github.com/matrix-org/matrix-react-sdk/pull/11256)). Fixes vector-im/element-web#3222.

### CODE REVIEW

File: src/utils/MultiInviter.ts [hunk 1]
```diff
@@ -170,6 +172,27 @@ export default class MultiInviter {
                     errcode: USER_ALREADY_INVITED,
                     error: "Member already invited",
                 });
+            } else if (member?.membership === "ban") {
```
Comment (richvdh) [inline src/utils/MultiInviter.ts:175]: slight concern with checking before attempting the invite is it is vulnerable to Element's state drifting out of sync with the server.  We could refuse to invite someone who isn't actually banned, etc.  It might have been nicer to prompt to unban if the initial attempt to invite fails.

But I guess we already have this problem with the other checks here, so :man_shrugging:

File: src/utils/MultiInviter.ts [hunk 2]
```diff
@@ -170,6 +172,27 @@ export default class MultiInviter {
...
+                let proceed = false;
+                const ourMember = room.getMember(this.matrixClient.getSafeUserId());
+                if (!!ourMember && room.currentState.hasSufficientPowerLevelFor("ban", ourMember.powerLevel)) {
```
Comment (richvdh) [inline src/utils/MultiInviter.ts:178]: it might be worth noting that having `ban` permission is not all that is required. Cf https://spec.matrix.org/v1.7/rooms/v10/#authorization-rules 4.5.{4,5}:

> If the sender’s power level is greater than or equal to the kick level, and the target user’s power level is less than the sender’s power level, allow. Otherwise, reject.

In other words: you also need power level for "kick", and to have a higher power level than the target.

I'm not sure if we need to care about these edge-cases: it's a material improvement as it stands. Just wanted to check you were aware.

**Review by richvdh — CHANGES_REQUESTED** [review]
Needs test coverage please

File: src/utils/MultiInviter.ts (hunk 2 above)
Comment (t3chguy) [reply src/utils/MultiInviter.ts:178]: > I'm not sure if we need to care about these edge-cases: it's a material improvement as it stands. Just wanted to check you were aware.

I was not as the C-S API spec is unclear, it just says 

> The caller must have the required power level in order to perform this operation.

Comment (richvdh) [reply src/utils/MultiInviter.ts:178]: >  The caller must have the required power level in order to perform this operation.

Well, that's correct, if unhelpful :)

Comment (t3chguy) [reply src/utils/MultiInviter.ts]: Oh for sure :D

**Comment by t3chguy** [comment]
image

Bang on

File: src/utils/MultiInviter.ts [hunk 3]
```diff
@@ -170,6 +172,32 @@ export default class MultiInviter {
                     errcode: USER_ALREADY_INVITED,
                     error: "Member already invited",
                 });
+            } else if (member?.membership === "ban") {
+                let proceed = false;
+                const ourMember = room.getMember(this.matrixClient.getSafeUserId());
```
Comment (richvdh) [inline src/utils/MultiInviter.ts:177]:
```suggestion
                // Check if we can unban the invitee.
                // See https://spec.matrix.org/v1.7/rooms/v10/#authorization-rules, particularly 4.5.3 and 4.5.4.
                const ourMember = room.getMember(this.matrixClient.getSafeUserId());
```

File: src/utils/MultiInviter.ts [hunk 4]
```diff
@@ -170,6 +172,32 @@ export default class MultiInviter {
...
+                if (
+                    !!ourMember &&
+                    member.powerLevel < ourMember.powerLevel &&
+                    room.currentState.hasSufficientPowerLevelFor("ban", ourMember.powerLevel) &&
+                    room.currentState.hasSufficientPowerLevelFor("kick", ourMember.powerLevel)
+                ) {
```
Comment (richvdh) [inline src/utils/MultiInviter.ts:183]: it might have been nice to pull this out to a `canUnbanUser(ourMember, member)` function, but whatever

**Review by richvdh — APPROVED** [review]
lgtm

File: src/utils/MultiInviter.ts (hunk 4 above)
Comment (t3chguy) [reply src/utils/MultiInviter.ts:185]: Ideally it'd have been a util in the js-sdk but 🤷
